### PR TITLE
feat: prune config

### PR DESCRIPTION
**Pull request by catngh**
**What does this PR do?**

- New commands: `$prune whitelist` and `$prune remove` to add and remove whitelisted role from being pruned
- Bot DM user after being kicked

image
image
image

1 user doesn't have role `@test` but he has a whitelisted role so no prune happens
image

### CODE REVIEW

File: src/adapters/config.ts [hunk 1]
```diff
@@ -937,6 +938,32 @@ class Config extends Fetcher {
       }
     )
   }
+
+  public async getExcludedRole(params: { guild_id: string }) {
+    return await this.jsonFetch<ResponseGetGuildPruneExcludeResponse>(
+      `${API_BASE_URL}/configs/whitelist-prune?guild_id=${params.guild_id}`
+    )
+  }
```
Comment (tuanddd) [inline src/adapters/config.ts:946]:
```suggestion
  public async getExcludedRole({ guild_id }: { guild_id: string }) {
    return await this.jsonFetch<ResponseGetGuildPruneExcludeResponse>(
      `${API_BASE_URL}/configs/whitelist-prune`,
      {
          query: {
              guild_id
          }
      }
    )
  }
```

File: src/commands/community/prune/inactive.ts [hunk 2]
```diff
@@ -12,10 +12,13 @@ export async function pruneInactiveExecute(i: ButtonInteraction) {
   ) {
     return
   }
+  if (!i.guild) return
```
Comment (tuanddd) [inline src/commands/community/prune/inactive.ts:15]: throw error nha, `GuildIdNotFoundError({})`

File: src/commands/community/prune/remove.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    await deleteWhitelist(id, msg.guild, msg.author)
+
+    return {
+      messageOptions: {
+        embeds: [
+          getSuccessEmbed({
+            msg,
+            title: "Successfully set!",
+            description: `<@&${id}> successfully whitelisted`,
```
Comment (tuanddd) [inline src/commands/community/prune/remove.ts:43]: check lại description nè em

File: src/commands/community/prune/without.ts [hunk 4]
```diff
@@ -12,30 +12,44 @@ import { Command } from "types/common"
...
     i.customId !== "confirm_prune_inactive" ||
+    !i.guild ||
     i.user.id !== "567326528216760320" //hnh
   ) {
     return
   }
+  let count = 0
+  const whitelistRole = await getExcludedRoles(i.guild)
+  const whitelistIds: string[] = []
+  whitelistRole.forEach((role) => whitelistIds.push(role.id))
```
Comment (tuanddd) [inline src/commands/community/prune/without.ts:31]:
```suggestion
  const whitelistIds: string[] = whitelistRole.map(r => r.id)
```

File: src/commands/community/prune/without.ts [hunk 5]
```diff
@@ -12,30 +12,44 @@ import { Command } from "types/common"
...
-    await mem.kick("Missing role")
+    if (
+      mem.roles.cache.hasAny(...whitelistIds) ||
+      mem.permissions.has("ADMINISTRATOR")
+    )
+      return
     await mem.send({
       content: `Sorry to say this but you haven't had a role yet, so we have to remove you from ${i.guild?.name}\nYou are welcome to join again: ${invite?.url}`,
     })
+    await mem.kick("Missing role")
```
Comment (tuanddd) [inline src/commands/community/prune/without.ts:43]: chắc cũng nên ghi thêm là miss role gì cho dễ trace

**Review by tuanddd — CHANGES_REQUESTED** [review]

**Review by tuanddd — APPROVED** [review]

**Review by trkhoi — APPROVED** [review]

**Pull request merged**